### PR TITLE
Log memory configuration when bridge starts

### LINKED ISSUE #4255: Log memory configuration on startup
- Physical memory
- Node.js heap size
- User actual settings
- Warn users setting Node.js memory higher than possible

### PR DESCRIPTION
**Pull request by saberduck**
Fixes #4255

### CODE REVIEW

File: sonar-plugin/sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/nodejs/NodeCommandBuilderImpl.java [hunk 1]
```diff
@@ -87,6 +87,7 @@ public NodeCommandBuilder configuration(Configuration configuration) {
   @Override
   public NodeCommandBuilder maxOldSpaceSize(int maxOldSpaceSize) {
     nodeJsArgs("--max-old-space-size=" + maxOldSpaceSize);
+    LOG.info("Configured Node.JS max old space {}.", maxOldSpaceSize);
```
Comment (yassin-kammoun-sonarsource) [inline sonar-plugin/sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/nodejs/NodeCommandBuilderImpl.java:90]:
```suggestion
    LOG.info("Configured Node.js max-old-space-size: {}.", maxOldSpaceSize);
```

File: sonar-plugin/sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/nodejs/NodeCommandTest.java [hunk 2]
```diff
@@ -165,7 +166,9 @@ void test_max_old_space_size_setting() throws IOException {
     command.waitFor();
     Map map = new Gson().fromJson(output.toString(), Map.class);
     double total_available_size = (double) map.get("total_available_size");
-    assertThat(total_available_size).isGreaterThan(2048 * 1000);
+    assertThat(logTester.logs(LoggerLevel.INFO))
+      .contains("Configured Node.JS max old space " + maxOldSpaceSize + ".");
```
Comment (yassin-kammoun-sonarsource) [inline sonar-plugin/sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/nodejs/NodeCommandTest.java:170]:
```suggestion
      .contains("Configured Node.js max-old-space-size: " + maxOldSpaceSize + ".");
```

File: sonar-plugin/sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/bridge/BridgeServerImpl.java [hunk 3]
```diff
@@ -258,7 +258,10 @@ private void initNodeCommand(
     context
       .config()
       .getInt(MAX_OLD_SPACE_SIZE_PROPERTY)
-      .ifPresent(nodeCommandBuilder::maxOldSpaceSize);
+      .ifPresent(maxSpace -> {
+        nodeCommandBuilder.maxOldSpaceSize(maxSpace);
+        LOG.info("Configured Node.JS heap size limit {}.", maxSpace);
```
[the remark this answers is not in the document]
Comment (yassin-kammoun-sonarsource) [reply sonar-plugin/sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/bridge/BridgeServerImpl.java:263]:
```suggestion
        LOG.info("Configured Node.JS heap size limit: {}.", maxSpace);
```

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/EslintBasedRulesTest.java [hunk 4]
```diff
@@ -256,6 +257,32 @@ void jsFileNamedAsTsFile() {
...
+    var build = getSonarScanner()
+      .setProjectKey(projectKey)
+      .setSourceEncoding("UTF-8")
+      .setSourceDirs(".")
+      .setProperty("sonar.javascript.node.maxspace", "500000")
+      .setProjectDir(projectDir);
+
+    var buildResult = orchestrator.executeBuild(build);
+    assertThat(buildResult.isSuccess()).isTrue();
+    assertThat(buildResult.getLogs()).contains("Configured Node.JS max old space 500000.");
```
Comment (yassin-kammoun-sonarsource) [inline its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/EslintBasedRulesTest.java:273]:
```suggestion
    assertThat(buildResult.getLogs()).contains("Configured Node.js max-old-space-size: 500000.");
```

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/EslintBasedRulesTest.java [hunk 5]
```diff
@@ -256,6 +257,32 @@ void jsFileNamedAsTsFile() {
...
+    var osMem = Pattern.compile(
+      ".*OS memory \\d+m\\. Node.JS heap size limit: \\d+m\\..*",
```
Comment (yassin-kammoun-sonarsource) [inline its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/EslintBasedRulesTest.java:275]:
```suggestion
      ".*OS memory: \\d+ MB\\. Node.js heap size limit: \\d+ MB\\..*",
```

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/EslintBasedRulesTest.java [hunk 6]
```diff
@@ -256,6 +257,32 @@ void jsFileNamedAsTsFile() {
...
+      Pattern.DOTALL
+    );
+    assertThat(buildResult.getLogs()).matches(osMem);
+    var warn = Pattern.compile(
+      "WARN: Node.JS heap size limit \\d+ is higher than available memory \\d+. Check your configuration of sonar.javascript.node.maxspace",
```
Comment (yassin-kammoun-sonarsource) [inline its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/EslintBasedRulesTest.java:280]:
```suggestion
      "WARN: Node.js heap size limit (\\d+ MB) is higher than available memory (\\d+ MB). Check your configuration of sonar.javascript.node.maxspace.",
```

**Review by yassin-kammoun-sonarsource — APPROVED** [review]
LGTM!

All my comments are mainly suggestions. Feel free to ignore them if you don't find them useful (I really won't mind).

File: packages/bridge/src/server.ts [hunk 7]
```diff
@@ -44,6 +46,17 @@ const MAX_REQUEST_SIZE = '50mb';
  */
 const SHUTDOWN_TIMEOUT = 15_000;
 
+function logMemoryConfiguration() {
+  const osMem = Math.floor(os.totalmem() / 1_000_000);
+  const heapSize = Math.floor(v8.getHeapStatistics().heap_size_limit / 1_000_000);
+  info(`OS memory ${osMem}m. Node.JS heap size limit: ${heapSize}m.`);
```
[the remark this answers is not in the document]
Comment (saberduck) [reply packages/bridge/src/server.ts:52]: I prefer info so it's visible by default

File: packages/bridge/src/server.ts [hunk 8]
```diff
@@ -44,6 +46,17 @@ const MAX_REQUEST_SIZE = '50mb';
  */
 const SHUTDOWN_TIMEOUT = 15_000;
 
+function logMemoryConfiguration() {
```
[the remark this answers is not in the document]
Comment (saberduck) [reply packages/bridge/src/server.ts:49]: I would prefer to do this in the next PR

[the remark this answers is not in the document]
Comment (yassin-kammoun-sonarsource) [reply packages/bridge/src/server.ts:49]: 👍

File: packages/bridge/src/server.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (yassin-kammoun-sonarsource) [reply packages/bridge/src/server.ts]: 👍

**Pull request merged**